identity-account-storage/src/types/key_location.rs

```diff
@@ -5,56 +5,60 @@ use core::fmt::Debug;
 use core::fmt::Display;
 use core::fmt::Formatter;
 use core::fmt::Result;
-use identity_core::common::Fragment;
+use identity_core::crypto::KeyType;
+use identity_did::verification::MethodData;
 use identity_did::verification::MethodType;
-use serde::Deserialize;
-use serde::Serialize;
-
-use crate::types::Generation;
+use identity_iota_core::document::IotaVerificationMethod;
+use rand::distributions::Alphanumeric;
+use rand::prelude::ThreadRng;
+use rand::Rng;
+use seahash::SeaHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
 
 /// The storage location of a verification method key.
-#[derive(Clone, Hash, PartialEq, Eq, Deserialize, Serialize)]
+#[derive(Clone, Hash, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[non_exhaustive]
 pub struct KeyLocation {
-  method: MethodType,
-  fragment: Fragment,
-  generation: Generation,
+  pub key_type: KeyType,
+  pub fragment: String,
+  pub key_hash: u64,
 }
 
 impl KeyLocation {
-  /// Creates a new `KeyLocation`.
-  pub fn new(method: MethodType, fragment: String, generation: Generation) -> Self {
-    Self {
-      method,
-      fragment: Fragment::new(fragment),
+  pub fn new(key_type: KeyType, fragment: String, method_data: &MethodData) -> Self {
+    let mut hasher = SeaHasher::new();
+    method_data.hash(&mut hasher);
+    let key_hash = hasher.finish();
 
-      generation,
+    Self {
+      key_type,
+      fragment,
+      key_hash,
     }
   }
 
-  /// Returns the method type of the key location.
-  pub fn method(&self) -> MethodType {
-    self.method
-  }
-
-  /// Returns the fragment name of the key location.
-  pub fn fragment(&self) -> &Fragment {
-    &self.fragment
-  }
-
-  /// Returns the fragment name of the key location.
-  pub fn fragment_name(&self) -> &str {
-    self.fragment.name()
-  }
+  /// Generates a random location for a key of the given [`KeyType`].
+  pub fn random(key_type: KeyType) -> Self {
+    let mut thread_rng: ThreadRng = rand::thread_rng();
+    let fragment: String = (&mut thread_rng)
+      .sample_iter(Alphanumeric)
+      .take(32)
+      .map(char::from)
+      .collect();
+    let key_hash: u64 = (&mut thread_rng).gen();
 
-  /// Returns the integration generation when this key was created.
-  pub fn generation(&self) -> Generation {
-    self.generation
+    Self {
+      key_type,
+      fragment,
+      key_hash,
+    }
   }
 }
 
 impl Display for KeyLocation {
   fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-    f.write_fmt(format_args!("({}:{}:{})", self.generation, self.fragment, self.method))
+    f.write_fmt(format_args!("({}:{})", self.fragment, self.key_hash))
   }
```

Comment:
Fun fact: one can have multiple methods in a DID Document with the same fragment but different fragments (could even be the same hash come to think of it). Not sure if this is sufficient but I don't have a better suggestion.

Reply:
If it's the same key hash, which is based on the public key, then it must be the same key material in storage, right? Then that only complicates key deletion. If we remove a verification method and mark the key in storage as deletable, then we also need to handle the case where the same key might be referenced by another verification method. It only makes that more complicated, though, it should still possible to figure out whether a key is still needed.

Reply:
Indeed, with the `key_hash` it would at worst just point to the same private key. I do wonder if it won't cause problems for other `Storage` implementations though, since there's no guarantee they use `KeyLocation::to_string` or compare them in the same way. It's a bit odd not to depend on the `KeyType` field but that's probably irrelevant.

Reply:
I think the `KeyType` is only needed as the indication for what type of key is at that location, but is not otherwise part of the location itself. It could also be removed from the `KeyLocation` and passed explicitly, but that's one more parameter for certain functions (`key_sign`, `key_public`).
We could additionally prepend the key location string with the string representation of the key type, though I think that shouldn't absolutely be required?

> there's no guarantee they use `KeyLocation::to_string`

I've added documentation to the `KeyLocation` type to explicitly state the expectations, so technically implementers can choose their own formatting, as long as it depends on the fragment and the hash.